src/ember-app.js

```diff
@@ -287,77 +111,203 @@ class EmberApp {
       fastbootInfo: fastbootInfo
     });
 
-    let destroyAppInstanceTimer;
-    if (parseInt(destroyAppInstanceInMs, 10) > 0) {
-      // start a timer to destroy the appInstance forcefully in the given ms.
-      // This is a failure mechanism so that node process doesn't get wedged if the `visit` never completes.
-      destroyAppInstanceTimer = setTimeout(function() {
-        if (instance && !result.instanceDestroyed) {
-          result.instanceDestroyed = true;
-          result.error = new Error('App instance was forcefully destroyed in ' + destroyAppInstanceInMs + 'ms');
-          instance.destroy();
-        }
-      }, destroyAppInstanceInMs);
-    }
-
-    let instance;
-    return this.visitRoute(path, fastbootInfo, bootOptions, result)
-      .then(appInstance => {
-        instance = appInstance;
+    let destroy;
+    return this.app.boot()
+      .then((app) => app.buildInstance())
+      .then((appInstance) => {
+        debug('building application instance');
+        fastbootInfo.register(appInstance);
+        destroy = destroyAppInstance(appInstance, result, destroyAppInstanceInMs);
+        return appInstance.boot(bootOptions);
+      })
+      .then((appInstance) => {
+        result.instanceBooted = true;
+        return appInstance.visit(path, bootOptions);
       })
-      .then(() => {
-        if (!disableShoebox) {
-          // if shoebox is not disabled, then create the shoebox and send API data
-          createShoebox(doc, fastbootInfo);
-        }
+      .then((appInstance) => waitForApp(appInstance))
```

Comment:
My two cents: I really feel `waitForApp` should be a private function of this class.